probe-rs/src/flashing/flasher.rs

```diff
@@ -862,67 +864,28 @@ impl<'p> ActiveFlasher<'p, Program> {
         );
 
         // Transfer the bytes to RAM.
-        let begin_data = self.buffer_address(0);
-        self.load_data(begin_data, bytes)?;
-
-        let error_code = self
-            .call_function_and_wait(
-                &Registers {
-                    pc: into_reg(self.flash_algorithm.pc_program_page)?,
-                    r0: Some(into_reg(address)?),
-                    r1: Some(bytes.len() as u32),
-                    r2: Some(into_reg(begin_data)?),
-                    r3: None,
-                },
-                false,
-                Duration::from_millis(
-                    self.flash_algorithm.flash_properties.program_page_timeout as u64,
-                ),
-            )
-            .map_err(|error| FlashError::PageWrite {
-                page_address: address,
-                source: Box::new(error),
-            })?;
-        tracing::info!("Flashing took: {:?}", t1.elapsed());
+        let begin_data = self.load_page_buffer(bytes, 0)?;
 
-        if error_code != 0 {
-            Err(FlashError::PageWrite {
-                page_address: address,
-                source: Box::new(FlashError::RoutineCallFailed {
-                    name: "program_page",
-                    error_code,
-                }),
-            })
-        } else {
-            self.progress.page_programmed(page.size(), t1.elapsed());
-            Ok(())
-        }
-    }
+        self.start_program_page_with_buffer(begin_data, address, bytes.len() as u32)?;
+        self.wait_for_write_end(address)?;
 
-    fn buffer_address(&self, buffer_number: usize) -> u64 {
-        // Ensure the buffer number is valid, otherwise there is a bug somewhere
-        // in the flashing code.
-        assert!(
-            buffer_number < self.flash_algorithm.page_buffers.len(),
-            "Trying to use non-existing buffer ({}/{}) for flashing. This is a bug. Please report it.",
-            buffer_number, self.flash_algorithm.page_buffers.len()
-        );
+        tracing::info!("Flashing took: {:?}", t1.elapsed());
 
-        self.flash_algorithm.page_buffers[buffer_number]
+        self.progress.page_programmed(page.size(), t1.elapsed());
+        Ok(())
     }
 
     pub(super) fn start_program_page_with_buffer(
         &mut self,
+        buffer_address: u64,
         page_address: u64,
-        buffer_number: usize,
+        data_size: u32,
```

Comment:
Why is this a `u32`? If this is a value in a register, u64 would fit better I think.

Reply:
`page.size()` is `u32` but I guess there's no reason not to widen the argument.